In [270]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [271]:
# Load the given datasets
df_categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
df_numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
print(df_categorical)
print(df_numerical)


     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
The datasets contain biometric information that can be used to predict whether or not a person has chronic kidney disease. The features are the different biometric information that are indicators of chronic kidney disease, which is the label.

Numerical dataset:
The features in this dataset are numerical. They are: age, blood pressure, blood glucose random, lood urea, serum creatinine, sodium, potassium, hemoglobin, packed cell volume, white blood cell count, red blood cell count. The label is the target column, indicating whether or not they have ckd. 


Categorical dataset:
The features in this dataset are categorical (nominal). They are: albumin, sugar, red blood cells, pus cell, pus cell clumps, bacteria, hypertension, diabetes mellitus, coronary artery disease, appetite, pedal edema, anemia. The label is the target column, indicating whether or not they have ckd. 

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [272]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
dup_numerical = df_numerical[df_numerical.duplicated(subset=['unique_id'])]
num_dup_numerical = dup_numerical.shape[0]
print(f"Number of duplicate rows in the numerical dataset: {num_dup_numerical}")

# Drop these duplicate rows
df_numerical_cleaned = df_numerical.drop_duplicates(subset=['unique_id'])

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
dup_categorical = df_categorical[df_categorical.duplicated(subset=['unique_id'])]
num_dup_categorical = dup_categorical.shape[0]
print(f"Number of duplicate rows in the categorical dataset: {num_dup_categorical}")
print(f"Duplicate rows in the categorical dataset: \n{dup_categorical}")
df_categorical_cleaned = df_categorical.drop_duplicates(subset=['unique_id'])


Number of duplicate rows in the numerical dataset: 30
Number of duplicate rows in the categorical dataset: 30
Duplicate rows in the categorical dataset: 
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
231     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
401     628271  2.0  1.0  abnormal    normal  notpresent  notpresent  yes   
402     158929  1.0  0.0       NaN       NaN  notpresent  notpresent   no   
403     240975  0.0  0.0    normal    normal  notpresent  notpresent   no   
404     441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
405     337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
406     118191  0.0  0.0    normal    normal  notpresent  notpresent   no   
407     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
408     302906  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
409     938027  0.0  0.0    normal    normal  notpresent  notpresent   no   

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: Not all unique ids are present in both datasets because the categorical dataset has more ids than the numerical dataset. The rows that are missing from one of the datasets will have NAN as their values for the columns from the other dataset, since there is no data to populate it with in the combined table. 

In [273]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
df_combined = pd.merge(df_numerical, df_categorical, on='unique_id')
#Print the combined dataset
df_combined

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
4,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
477,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
478,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
479,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
480,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [274]:
# Calculate the percentage of rows that contain atleast one missing value
missing_rows = df_combined.isnull().any(axis=1).sum()
total_rows = df_combined.shape[0]
missing_percentage = (missing_rows/total_rows) * 100
# Print %
print(f"The missing percentage is: {missing_percentage}%")
# Drop these rows from the dataset
df_combined = df_combined.dropna()
# Print the Dataset
df_combined

The missing percentage is: 58.921161825726145%


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
467,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
469,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
478,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
479,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [275]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
df_sorted = df_combined.sort_values(by='Target')
df_sorted.reset_index(drop=True, inplace=True)

# Print the dataset
df_sorted

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
4,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
193,769512,38.0,60.0,91.0,36.0,0.7,135.0,3.7,14.0,46.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
194,769512,38.0,60.0,91.0,36.0,0.7,135.0,3.7,14.0,46.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
195,261137,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
196,489160,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [276]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
df_encoded = pd.get_dummies(df_sorted, columns=categorical_columns)

# Print the dataset
df_encoded

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,False,True,False,True,True,False,True,False,True,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,True,False,False,True,True,False,False,True,True,False
3,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,True,False,False,True,False,True,False,True,True,False
4,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,True,False,False,True,False,True,False,True,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
193,769512,38.0,60.0,91.0,36.0,0.7,135.0,3.7,14.0,46.0,...,True,False,True,False,True,False,True,False,False,True
194,769512,38.0,60.0,91.0,36.0,0.7,135.0,3.7,14.0,46.0,...,True,False,True,False,True,False,True,False,False,True
195,261137,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,...,True,False,True,False,True,False,True,False,False,True
196,489160,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,...,True,False,True,False,True,False,True,False,False,True


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [277]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in df_encoded.columns:
    mean = df_encoded[column].mean()
    std = df_encoded[column].std()        
    upper_bound = mean + (3 * std)
    lower_bound = mean - (3 * std)        
    df_cleaned = df_encoded[(df_encoded[column] >= lower_bound) & (df_encoded[column] <= upper_bound)]

df_cleaned.reset_index(drop=True, inplace=True)

# Print the dataset
df_cleaned

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,False,True,False,True,True,False,True,False,True,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,True,False,False,True,True,False,False,True,True,False
3,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,True,False,False,True,False,True,False,True,True,False
4,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,True,False,False,True,False,True,False,True,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
193,769512,38.0,60.0,91.0,36.0,0.7,135.0,3.7,14.0,46.0,...,True,False,True,False,True,False,True,False,False,True
194,769512,38.0,60.0,91.0,36.0,0.7,135.0,3.7,14.0,46.0,...,True,False,True,False,True,False,True,False,False,True
195,261137,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,...,True,False,True,False,True,False,True,False,False,True
196,489160,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,...,True,False,True,False,True,False,True,False,False,True


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [278]:
# Normalize the all Numerical Attributes in the dataset.
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df_cleaned[numerical_columns] = scaler.fit_transform(df_cleaned[numerical_columns])

# Print the dataset
df_cleaned

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,0.194805,0.666667,0.088095,0.100334,0.087838,0.358974,0.022472,0.353741,0.311111,...,True,False,True,False,True,False,False,True,True,False
1,137148,0.870130,0.833333,0.535714,0.267559,0.351351,0.743590,0.008989,0.414966,0.466667,...,False,True,False,True,True,False,True,False,True,False
2,484175,0.610390,0.666667,0.000000,0.324415,0.459459,0.076923,0.026966,0.435374,0.444444,...,True,False,False,True,True,False,False,True,True,False
3,337834,0.545455,0.333333,0.111905,0.153846,0.229730,0.000000,0.000000,0.551020,0.511111,...,True,False,False,True,False,True,False,True,True,False
4,337834,0.545455,0.333333,0.111905,0.153846,0.229730,0.000000,0.000000,0.551020,0.511111,...,True,False,False,True,False,True,False,True,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
193,769512,0.415584,0.166667,0.050000,0.086957,0.020270,0.615385,0.026966,0.741497,0.822222,...,True,False,True,False,True,False,True,False,False,True
194,769512,0.415584,0.166667,0.050000,0.086957,0.020270,0.615385,0.026966,0.741497,0.822222,...,True,False,True,False,True,False,True,False,False,True
195,261137,0.376623,0.166667,0.083333,0.096990,0.006757,0.615385,0.031461,0.789116,0.755556,...,True,False,True,False,True,False,True,False,False,True
196,489160,0.285714,0.166667,0.021429,0.133779,0.006757,0.871795,0.056180,0.986395,0.933333,...,True,False,True,False,True,False,True,False,False,True


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:
The unique_id column is not very important for the modeling and predictions since it doesn't reveal any appropriate information for predictions, it is simply the unique id number referencing the particular combination of biometric information. 

In [279]:
#Remove that column
df_cleaned = df_cleaned.drop(['unique_id'], axis=1)

# Print the dataset
df_cleaned

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.194805,0.666667,0.088095,0.100334,0.087838,0.358974,0.022472,0.353741,0.311111,0.380531,...,True,False,True,False,True,False,False,True,True,False
1,0.870130,0.833333,0.535714,0.267559,0.351351,0.743590,0.008989,0.414966,0.466667,0.141593,...,False,True,False,True,True,False,True,False,True,False
2,0.610390,0.666667,0.000000,0.324415,0.459459,0.076923,0.026966,0.435374,0.444444,0.367257,...,True,False,False,True,True,False,False,True,True,False
3,0.545455,0.333333,0.111905,0.153846,0.229730,0.000000,0.000000,0.551020,0.511111,0.128319,...,True,False,False,True,False,True,False,True,True,False
4,0.545455,0.333333,0.111905,0.153846,0.229730,0.000000,0.000000,0.551020,0.511111,0.128319,...,True,False,False,True,False,True,False,True,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
193,0.415584,0.166667,0.050000,0.086957,0.020270,0.615385,0.026966,0.741497,0.822222,0.234513,...,True,False,True,False,True,False,True,False,False,True
194,0.415584,0.166667,0.050000,0.086957,0.020270,0.615385,0.026966,0.741497,0.822222,0.234513,...,True,False,True,False,True,False,True,False,False,True
195,0.376623,0.166667,0.083333,0.096990,0.006757,0.615385,0.031461,0.789116,0.755556,0.088496,...,True,False,True,False,True,False,True,False,False,True
196,0.285714,0.166667,0.021429,0.133779,0.006757,0.871795,0.056180,0.986395,0.933333,0.119469,...,True,False,True,False,True,False,True,False,False,True


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [280]:
# Export the dataframe to a new csv file
df_cleaned.to_csv('cleaned_data.csv', index=False)

# Export the dataframe to a new json file
df_cleaned.to_json('cleaned_data.json', orient='records')


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [281]:
df_cleaned.head(15)

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.194805,0.666667,0.088095,0.100334,0.087838,0.358974,0.022472,0.353741,0.311111,0.380531,...,True,False,True,False,True,False,False,True,True,False
1,0.870130,0.833333,0.535714,0.267559,0.351351,0.743590,0.008989,0.414966,0.466667,0.141593,...,False,True,False,True,True,False,True,False,True,False
2,0.610390,0.666667,0.000000,0.324415,0.459459,0.076923,0.026966,0.435374,0.444444,0.367257,...,True,False,False,True,True,False,False,True,True,False
3,0.545455,0.333333,0.111905,0.153846,0.229730,0.000000,0.000000,0.551020,0.511111,0.128319,...,True,False,False,True,False,True,False,True,True,False
4,0.545455,0.333333,0.111905,0.153846,0.229730,0.000000,0.000000,0.551020,0.511111,0.128319,...,True,False,False,True,False,True,False,True,True,False
5,0.714286,0.500000,0.245238,0.461538,0.236486,0.615385,0.060674,0.312925,0.333333,0.238938,...,False,True,False,True,False,True,False,True,True,False
6,0.636364,0.666667,0.483333,0.752508,0.932432,0.538462,0.020225,0.353741,0.288889,0.477876,...,True,False,False,True,False,True,False,True,True,False
7,0.649351,0.666667,0.140476,0.324415,0.425676,0.512821,0.051685,0.408163,0.444444,0.115044,...,True,False,True,False,True,False,True,False,True,False
8,0.701299,0.666667,0.083333,0.143813,0.128378,0.641026,0.060674,0.544218,0.533333,0.296460,...,True,False,True,False,True,False,True,False,True,False
9,0.558442,0.833333,0.140476,0.494983,0.770270,0.282051,0.015730,0.340136,0.333333,0.256637,...,True,False,False,True,False,True,False,True,True,False


### Paste the markdown table here
|    |      age |    bp |      bgr |      bu |        sc |      sod |      pot |     hemo |       pcv |      wbcc | ... | pc_normal | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_poor | pe_yes | ane_yes | Target_notckd |
|----|----------|-------|----------|---------|-----------|----------|----------|----------|-----------|-----------|-----|-----------|-------------|------------|---------|--------|---------|------------|--------|---------|----------------|
|  0 | 0.897059 |  1.00 | 0.965665 | 0.606061| 0.825397  | 0.666667 | 0.000000 | 0.011494 | 0.076923  | 0.247706  | ... | False     | True        | False      | True    | True   | True    | True       | False  | False   | False          |
|  1 | 0.647059 | 0.75  | 0.253219 | 0.734848| 1.000000  | 0.366667 | 0.655172 | 0.000000 | 0.038462  | 0.192661  | ... | False     | False       | False      | True    | False  | False   | False       | False  | False   | False          |
|  2 | 0.705882 | 0.75  | 0.150215 | 0.325758| 0.301587  | 0.533333 | 0.793103 | 0.229885 | 0.192308  | 0.568807  | ... | True      | False       | False      | False   | False  | False   | False       | False  | False   | False          |
|  3 | 0.500000 | 0.00  | 0.399142 | 0.621212| 0.460317  | 0.700000 | 0.379310 | 0.080460 | 0.000000  | 0.944954  | ... | True      | False       | False      | True    | True   | False   | False       | False  | False   | False          |
|  4 | 0.691176 | 0.50  | 1.000000 | 0.189394| 0.142857  | 0.066667 | 0.206897 | 0.149425 | 0.269231  | 0.605505  | ... | True      | False       | False      | False   | True   | False   | True        | False  | False   | False          |
|  5 | 0.750000 | 1.00  | 0.901288 | 0.189394| 0.444444  | 0.766667 | 0.206897 | 0.448276 | 0.461538  | 0.504587  | ... | True      | False       | True       | True    | False  | True    | False       | False  | False   | False          |
|  6 | 0.897059 | 0.50  | 0.785408 | 1.000000| 0.666667  | 0.600000 | 1.000000 | 0.160920 | 0.192308  | 0.266055  | ... | False     | False       | False      | True    | True   | True    | False       | False  | False   | False          |
|  7 | 0.764706 | 0.00  | 0.725322 | 0.363636| 0.619048  | 0.566667 | 0.862069 | 0.045977 | 0.038462  | 0.293578  | ... | False     | False       | True       | True    | True   | False   | True        | True   | False   | False          |
|  8 | 0.632353 | 0.50  | 0.618026 | 0.477273| 0.555556  | 0.566667 | 0.689655 | 0.206897 | 0.230769  | 0.284404  | ... | False     | True        | True       | True    | True   | False   | False       | False  | False   | False          |
|  9 | 0.838235 | 0.25  | 0.618026 | 0.651515| 0.936508  | 0.000000 | 0.344828 | 0.034483 | 0.000000  | 0.660550  | ... | False     | True        | True       | True    | True   | True    | True        | True   | False   | False          |
| 10 | 0.867647 | 0.00  | 0.206009 | 0.871212| 0.777778  | 0.533333 | 0.689655 | 0.264368 | 0.269231  | 1.000000  | ... | True      | False       | False      | True    | True   | False   | True        | True   | False   | False          |
| 11 | 0.867647 | 0.25  | 0.639485 | 0.545455| 0.507937  | 0.433333 | 0.517241 | 0.149425 | 0.192308  | 0.119266  | ... | False     | True        | True       | True    | True   | True    | True        | True   | False   | False          |
| 12 | 0.647059 | 0.25  | 0.600858 | 0.121212| 0.206349  | 0.533333 | 0.310345 | 0.804598 | 0.923077  | 0.752294  | ... | True      | False       | False      | False   | False  | False   | False       | False  | False   | False          |
| 13 | 0.838235 | 0.25  | 0.832618 | 0.583333| 0.365079  | 0.333333 | 0.379310 | 0.390805 | 0.346154  | 0.486239  | ... | False     | False       | False      | True    | True   | True    | False       | True   | False   | False          |
| 14 | 0.735294 | 0.25  | 0.223176 | 0.242424| 0.206349  | 0.533333 | 0.620690 | 0.402299 | 0.423077  | 0.330275  | ... | False     | False       | False      | True    | True   | False   | False       | False  | False   | False          |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: We can mitigate data conversion hallucinations from LLM conversions by using smaller chunks of data to avoid the hallucinations. It can also be helpful to have regular human checks to validate the conversions and ensure that the converted data is accurate.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [282]:
# Code to rename all the columns in the dataset

# Define the mapping of abbreviations to full names
rename_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc_abnormal': 'red blood cells abnormal',
    'rbc_normal': 'red blood cells normal',
    'pc_abnormal': 'pus cell abnormal',
    'pc_normal': 'pus cell normal',
    'pcc_notpresent': 'pus cell clumps not present',
    'pcc_present': 'pus cell clumps present',
    'ba_notpresent': 'bacteria not present',
    'ba_present': 'bacteria present',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rbcc': 'red blood cell count',
    'htn_no': 'hypertension no',
    'htn_yes': 'hypertension yes',
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_no': 'coronary artery disease no',
    'cad_yes': 'coronary artery disease yes',
    'appet_good': 'appetite good',
    'appet_poor': 'appetite poor',
    'pe_no': 'pedal edema no',
    'pe_yes': 'pedal edema yes',
    'ane_no': 'anemia no',
    'ane_yes': 'anemia yes',
    'Target_ckd': 'ckd yes',
    'Target_notckd': 'ckd no'
}

df_cleaned.rename(columns=rename_mapping, inplace=True)
df_cleaned.columns

Index(['age', 'blood pressure', 'blood glucose random', 'blood urea',
       'serum creatinine', 'sodium', 'potassium', 'hemoglobin',
       'packed cell volume', 'white blood cell count', 'red blood cell count',
       'albumin', 'sugar', 'red blood cells abnormal',
       'red blood cells normal', 'pus cell abnormal', 'pus cell normal',
       'pus cell clumps not present', 'pus cell clumps present',
       'bacteria not present', 'bacteria present', 'hypertension no',
       'hypertension yes', 'diabetes mellitus no', 'diabetes mellitus yes',
       'coronary artery disease no', 'coronary artery disease yes',
       'appetite good', 'appetite poor', 'pedal edema no', 'pedal edema yes',
       'anemia no', 'anemia yes', 'ckd yes', 'ckd no'],
      dtype='object')

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [283]:
# Converted SQL to Pandas code
result = df_cleaned.groupby('ckd yes').size().reset_index(name='count')
print(result)

   ckd yes  count
0    False    150
1     True     48


*Describe what the above code does here*

The code above groups the rows in the dataframe by the unique values in the 'ckd yes' column. Each unique value in this column will form a group. It then calculates the size of each group using the .size() function, and creates a pandas series where the index is the unique values from the ckd yes column and the values are the counts of the occurences of each value. Then, it converts the resulting series into a dataframe, resets the index into default integer indexing, and creates a new column called "count" that has the counts of the unique values. 
Essentially, there are 48 rows of metrics that indicated ckd and 150 rows that did not have ckd.